### ISSUE TITLE
Error when using pytorch in Notebooks

### ISSUE THREAD

**Issue by hno2** [kubeflow/kubeflow#2522]
Hey, 

when using pytorch (via fastai) in my notebook I get the error:
 ````
RuntimeError: Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 138, in _worker_loop
    samples = collate_fn([dataset[i] for i in batch_indices])
  File "/opt/conda/lib/python3.6/site-packages/fastai/torch_core.py", line 117, in data_collate
    return torch.utils.data.dataloader.default_collate(to_data(batch))
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 232, in default_collate
    return [default_collate(samples) for samples in transposed]
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 232, in <listcomp>
    return [default_collate(samples) for samples in transposed]
  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 207, in default_collate
    storage = batch[0].storage()._new_shared(numel)
  File "/opt/conda/lib/python3.6/site-packages/torch/storage.py", line 122, in _new_shared
    return cls._new_using_fd(size)
RuntimeError: unable to write to file </torch_1509_1969058146>
````
I read online that this might be solved by setting `--ipc=host` for docker. Where can I set this? Or can this error be solved in another way?

**Comment by jlewi** [comment]
Do you have more information about what pytorch is trying to do? e.g. is it trying to write to a file? Is it trying to talk to other hosts in the PyTorch job?

Without knowing more its hard to know what the correct solution is.

`--ipc=host` is a networking setting for docker. I think we need to know more about what pytorch is trying to do in order to map it onto Kubernetes & Kubeflow.

/cc @johnugeorge @richardsliu

**Comment by hno2** [comment]
I'm not sure, how or if I can provide you with the information you requested, since I'm just getting started with both kubeflow and pytorch. 

I will still share what i found out, so far. 
My Information about `--ipc=host` came from the [docu](https://github.com/pytorch/pytorch/blob/master/README.md#docker-image) of pytorch:

> Please note that PyTorch uses shared memory to share data between processes, so if torch multiprocessing is used (e.g. for multithreaded data loaders) the default shared memory segment size that container runs with is not enough, and you should increase shared memory size either with `--ipc=hos`t or `--shm-size` command line options to `nvidia-docker run`.

My understanding is, that fastai makes strong use of torch multiprocessing. At least for my `unable to write to file </torch_....>`-Error, this seems to be the case.


This error occurs for both[` fastai.fit_one_cycle()`](https://github.com/fastai/fastai/blob/master/fastai/train.py#L14) and [`DataBunch.show_batch()`](https://github.com/fastai/fastai/blob/master/fastai/basic_data.py#L183). The second one fails when trying to get a shared storage from pytorch.multiprocessing.

Is there any way I can set `--shm-size` while generating the notebook container?

**Comment by johnugeorge** [comment]
Can you provide more info about the env? 
How did you run the PyTorchJob?

**Comment by hno2** [comment]
I used downloaded [this notebook](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson1-pets.ipynb).
I used the tensorflow 1.12.0-GPU Notebook-Baseimage and installed fastai directly via the notebook.

**Comment by hno2** [comment]
Any news ?
Is there a way to set `--shm-size` somewhere?

**Comment by pdmack** [comment]
Sorry, just seeing this. I'll have a look.

**Comment by dmrub** [comment]
We had similar problem and found this solution: https://github.com/dmrub/kubeflow/commit/9b31f7ee866b793ee1fbf79354ac7f137eb7adb2, I modified Jupyter Spawner configuration in my Kubeflow fork: https://github.com/dmrub/kubeflow/commit/9b31f7ee866b793ee1fbf79354ac7f137eb7adb2

**Comment by hno2** [comment]
I just upgraded to 0.5 hoping adding this was easier in this version. How can I add shm or mount` /dev/shm` with the new juypter spawner?

**Comment by hno2** [comment]
Ok thank you for pointing me in the right direction. I edited this file, to include the empty dir `dshm `as described by @dmrub.
 https://github.com/kubeflow/kubeflow/blob/master/components/jupyter-web-app/kubeflow_jupyter/common/utils.py#L93-L104 

Now fastai and pythorch are working without a issue.

@pdmack  @johnugeorge Do you want me to add a pull request adding a option for this in the jupyter-web-ui?

**Comment by pdmack** [comment]
Sounds great to me.

**Comment by hno2** [comment]
fyi, I added this as pull request #3102 

I will close this Issue once it is merged.

**Comment by jlewi** [comment]
@hno2 Thanks for figuring this out! I was going through the fast ai course and hit this issue.

While we wait for #3102 to get merged here's a potential work around.

1. Spawn a notebook using the notebook manager UI
1. Get the spec for the notebook

    ```
      kubectl get notebooks -o yaml <notebook>  > /tmp/mynotebook.yaml
    ```

1. Edit the notebook spec

   * Add to volumes
     
      ```
      - name: dshm
        emptyDir:
          medium: Memory
      ```

   * Add to volume mounts

      ```
     - mountPath: /dev/shm
        name: dshm
      ```

1. Apply your notebook
     
     ```
     kubectl apply -f  / /tmp/mynotebook.yaml
     ```

**Comment by jlewi** [comment]
Before we close this issue out we should

1. Build a new image with the fix
1. Cherry pick it onto the 0.5 branch.

**Comment by hno2** [comment]
@jlewi #3102 is merged now. Will you take it from here ?

**Comment by hno2** [comment]
@jlewi Not sure how your release process works, but if these changes where included in v0.5.1 - which was released today - i guess this issue can be closed?

**Comment by jlewi** [comment]
@hno2 this is in 0.6 which we recently released.

Here's a screenshot of the option to enable SHM.

![shm](https://user-images.githubusercontent.com/777219/62828277-de440900-bb96-11e9-8481-c2a83e3cf7ac.png)
